File: app/src/main/java/li/klass/fhem/fragments/ConnectionDetailFragment.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.fragments;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.ResultReceiver;
import android.support.v4.app.FragmentActivity;
import android.text.method.PasswordTransformationMethod;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import java.io.File;
import java.io.Serializable;
import java.util.Arrays;
import java.util.List;

import li.klass.fhem.R;
import li.klass.fhem.constants.Actions;
import li.klass.fhem.constants.BundleExtraKeys;
import li.klass.fhem.constants.ResultCodes;
import li.klass.fhem.fhem.connection.FHEMServerSpec;
import li.klass.fhem.fhem.connection.ServerType;
import li.klass.fhem.fragments.core.BaseFragment;
import li.klass.fhem.service.intent.ConnectionsIntentService;
import li.klass.fhem.ui.FileDialog;
import li.klass.fhem.util.DialogUtil;

import static com.google.common.collect.Lists.newArrayList;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION_CLIENT_CERTIFICATE_PASSWORD;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION_CLIENT_CERTIFICATE_PATH;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION_ENABLE_CLIENT_CERTIFICATE;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION_SERVER_CERTIFICATE_PATH;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION_URL;
import static li.klass.fhem.constants.BundleExtraKeys.CONNECTION_USERNAME;
import static li.klass.fhem.constants.ResultCodes.SUCCESS;

public class ConnectionDetailFragment extends BaseFragment {

    private static final String TAG = ConnectionDetailFragment.class.getName();
    private String connectionId;
    private boolean isModify = false;
    private ServerType connectionType;
    private ConnectionTypeDetailChangedListener detailChangedListener = null;

    @Override
    public void setArguments(Bundle args) {
        super.setArguments(args);
        if (args.containsKey(BundleExtraKeys.CONNECTION_ID)) {
            connectionId = args.getString(BundleExtraKeys.CONNECTION_ID);
            isModify = true;
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        if (view != null) {
            return view;
        }

        view = inflater.inflate(R.layout.connection_detail, container, false);
        assert (view != null);

        Spinner connectionTypeSpinner = (Spinner) view.findViewById(R.id.connectionType);
        if (isModify) {
            connectionTypeSpinner.setEnabled(false);
        }

        final List<ServerType> connectionTypes = getServerTypes();

        ArrayAdapter<ServerType> adapter = new ArrayAdapter<>(getActivity(),
                android.R.layout.simple_spinner_dropdown_item, connectionTypes);
        connectionTypeSpinner.setAdapter(adapter);

        connectionTypeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int position, long id) {
                handleConnectionTypeChange(connectionTypes.get(position));
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
            }
        });

        Button saveButton = (Button) view.findViewById(R.id.save);
        saveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                handleSave();
            }
        });

        return view;
    }

    private List<ServerType> getServerTypes() {
        final List<ServerType> connectionTypes = newArrayList();
        connectionTypes.addAll(Arrays.asList(ServerType.values()));
        connectionTypes.remove(ServerType.DUMMY);
        return connectionTypes;
    }

    @SuppressLint("InflateParams")
    private void handleConnectionTypeChange(ServerType connectionType) {
        if (getView() == null) return;

        this.connectionType = connectionType;
        LayoutInflater inflater = getActivity().getLayoutInflater();

        View view;
        if (connectionType == ServerType.FHEMWEB) {
            view = inflater.inflate(R.layout.connection_fhemweb, null);
            handleFHEMWEBView(view);
        } else if (connectionType == ServerType.TELNET) {
            view = inflater.inflate(R.layout.connection_telnet, null);
        } else {
            throw new IllegalArgumentException("cannot handle connection type " + connectionType);
        }

        assert view != null;

        CheckBox showPasswordCheckbox = (CheckBox) view.findViewById(R.id.showPasswordCheckbox);
        final EditText passwordView = (EditText) view.findViewById(R.id.password);
        if (showPasswordCheckbox != null && passwordView != null) {
            showPasswordCheckbox.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    CheckBox radio = (CheckBox) view;
                    boolean checked = radio.isChecked();
                    if (checked) {
                        passwordView.setTransformationMethod(null);
                    } else {
                        passwordView.setTransformationMethod(PasswordTransformationMethod.getInstance());
                    }
                }
            });
        }

        LinearLayout connectionPreferences = (LinearLayout) getView().findViewById(R.id.connectionPreferences);
        connectionPreferences.removeAllViews();
        connectionPreferences.addView(view);

        if (detailChangedListener != null) detailChangedListener.onChanged();
    }

    private void handleSave() {
        Intent intent = new Intent();
        intent.setClass(getActivity(), ConnectionsIntentService.class);
        if (isModify) {
            intent.setAction(Actions.CONNECTION_UPDATE);
            intent.putExtra(BundleExtraKeys.CONNECTION_ID, connectionId);
        } else {
            intent.setAction(Actions.CONNECTION_CREATE);
        }

        intent.putExtra(BundleExtraKeys.CONNECTION_TYPE, connectionType.name());

        String name = getTextViewContent(R.id.name);
        if (enforceNotEmpty(R.string.connectionName, name)) return;
        intent.putExtra(BundleExtraKeys.CONNECTION_NAME, name);

        intent.putExtra(BundleExtraKeys.CONNECTION_PASSWORD, getTextViewContent(R.id.password));

        switch (connectionType) {
            case TELNET:
                if (!handleTelnetSave(intent)) {
                    return;
                }
                break;

            case FHEMWEB:
                if (!handleFHEMWEBSave(intent)) {
                    return;
                }
                break;
        }

        intent.putExtra(BundleExtraKeys.RESULT_RECEIVER, new ResultReceiver(new Handler()) {
            @Override
            protected void onReceiveResult(int resultCode, Bundle resultData) {
                super.onReceiveResult(resultCode, resultData);

                if (resultCode != ResultCodes.SUCCESS) {
                    Log.e(TAG, "could not save! resultCode=" + resultCode + ",resultData=" + resultData);
                    return;
                }

                Intent intent = new Intent(Actions.BACK);
                getActivity().sendBroadcast(intent);
            }
        });

        getActivity().startService(intent);
    }

    private void handleFHEMWEBView(View view) {
        Button setClientCertificate = (Button) view.findViewById(R.id.setClientCertificatePath);
        setClientCertificate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (getView() == null) return;

                final TextView clientCertificatePath = (TextView) getView().findViewById(R.id.clientCertificatePath);
                File initialPath = new File(clientCertificatePath.getText().toString());

                FileDialog fileDialog = new FileDialog(view.getContext(), initialPath);
                fileDialog.addFileListener(new FileDialog.FileSelectedListener() {
                    @Override
                    public void fileSelected(File file) {
                        clientCertificatePath.setText(file.getAbsolutePath());
                    }
                });
                fileDialog.showDialog();
            }
        });

        Button setServerCertificate = (Button) view.findViewById(R.id.setServerCertificatePath);
        setServerCertificate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (getView() == null) return;

                final TextView serverCertificatePath = (TextView) getView().findViewById(R.id.serverCertificatePath);
                File initialPath = new File(serverCertificatePath.getText().toString());

                FileDialog fileDialog = new FileDialog(view.getContext(), initialPath);
                fileDialog.addFileListener(new FileDialog.FileSelectedListener() {
                    @Override
                    public void fileSelected(File file) {
                        serverCertificatePath.setText(file.getAbsolutePath());
                    }
                });
                fileDialog.showDialog();
            }
        });
    }

    private String getTextViewContent(int id) {
        View view = getView();
        if (view == null) return null;

        TextView textView = (TextView) view.findViewById(id);
        if (textView == null) {
            Log.e(TAG, "cannot find " + id);
            return null;
        }
        return String.valueOf(textView.getText());
    }

    private boolean enforceNotEmpty(int fieldName, String value) {
        if (value != null && value.trim().length() > 0) {
            return false;
        }

        Context context = getActivity();
        String emptyError = context.getString(R.string.connectionEmptyError);
        String errorMessage = String.format(emptyError, context.getString(fieldName));

        DialogUtil.showAlertDialog(context, R.string.error, errorMessage);

        return true;
    }

    private boolean handleTelnetSave(Intent intent) {
        String ip = getTextViewContent(R.id.ip);
        if (enforceNotEmpty(R.string.connectionIP, ip)) return false;
        intent.putExtra(BundleExtraKeys.CONNECTION_IP, ip);

        String port = getTextViewContent(R.id.port);
        if (enforceNotEmpty(R.string.connectionPort, port)) return false;
        intent.putExtra(BundleExtraKeys.CONNECTION_PORT, port);

        return true;
    }

    private boolean handleFHEMWEBSave(Intent intent) {
        if (getView() == null) return false;

        String url = getTextViewContent(R.id.url);
        if (enforceNotEmpty(R.string.connectionURL, url)) return false;
        if (enforceUrlStartsWithHttp(url)) return false;

        intent.putExtra(CONNECTION_URL, url);

        String username = getTextViewContent(R.id.username);
        intent.putExtra(CONNECTION_USERNAME, username);

        String clientCertificatePath = getTextViewContent(R.id.clientCertificatePath);
        String serverCertificatePath = getTextViewContent(R.id.serverCertificatePath);

        CheckBox clientCertificateCheckbox = (CheckBox) getView().findViewById(R.id.enableCertificateAuthentication);
        boolean clientCertificateEnabled = clientCertificateCheckbox.isChecked();

        intent.putExtra(CONNECTION_CLIENT_CERTIFICATE_PATH, clientCertificatePath);
        intent.putExtra(CONNECTION_SERVER_CERTIFICATE_PATH, serverCertificatePath);
        intent.putExtra(CONNECTION_ENABLE_CLIENT_CERTIFICATE, clientCertificateEnabled);
        intent.putExtra(CONNECTION_CLIENT_CERTIFICATE_PASSWORD, getTextViewContent(R.id.clientCertificatePassword));

        return true;
    }

    private boolean enforceUrlStartsWithHttp(String url) {
        if (!url.startsWith("http")) {
            Context context = getActivity();
            String emptyError = context.getString(R.string.connectionUrlHttp);

            DialogUtil.showAlertDialog(context, R.string.error, emptyError);

            return true;
        }
        return false;
    }

    @Override
    public void update(boolean doUpdate) {
        if (!isModify) {
            Log.e(TAG, "I can only update if a connection is being modified!");
            getActivity().sendBroadcast(new Intent(Actions.DISMISS_EXECUTING_DIALOG));
            return;
        }

        Intent intent = new Intent(Actions.CONNECTION_GET);
        intent.setClass(getActivity(), ConnectionsIntentService.class);
        intent.putExtra(BundleExtraKeys.CONNECTION_ID, connectionId);
        intent.putExtra(BundleExtraKeys.RESULT_RECEIVER, new ResultReceiver(new Handler()) {

            @Override
            protected void onReceiveResult(int resultCode, Bundle resultData) {
                if (resultCode != SUCCESS || !resultData.containsKey(CONNECTION)) {
                    return;
                }

                Serializable serializable = resultData.getSerializable(CONNECTION);
                if (!(serializable instanceof FHEMServerSpec)) {
                    Log.e(TAG, "expected an FHEMServerSpec, but got " + serializable);
                    return;
                }

                setValuesForCurrentConnection((FHEMServerSpec) serializable);

                FragmentActivity activity = getActivity();
                if (activity != null)
                    activity.sendBroadcast(new Intent(Actions.DISMISS_EXECUTING_DIALOG));
            }
        });
        getActivity().startService(intent);
    }

    private void setValuesForCurrentConnection(final FHEMServerSpec connection) {
        final View view = getView();
        if (view == null) return;

        // We do not need to change the type selector here, as the right one is already selected.
        // We just overwrite values within the edit fields.
        if (connection.getServerType() == connectionType) {
            fillDetail(connection);
        } else {
            // We have to change the detail view to the one which is right for the current
            // connection type. However, we do not know when the selection changed listener
            // of the combo box fires. This is why we register a global listener, which is called
            // when the new view has been attached to the root view.
            // Afterwards we can continue with filling the fields with the respective values
            // of the current connection!
            detailChangedListener = new ConnectionTypeDetailChangedListener() {
                @Override
                public void onChanged() {
                    detailChangedListener = null;

                    fillDetail(connection);
                }
            };
            Spinner connectionTypeSpinner = (Spinner) view.findViewById(R.id.connectionType);
            connectionTypeSpinner.setSelection(selectionIndexFor(connection.getServerType()), true);
        }
    }

    private void fillDetail(FHEMServerSpec connection) {
        setTextViewContent(R.id.name, connection.getName());

        switch (connectionType) {
            case FHEMWEB:
                fillFHEMWEB(connection);
                break;
            case TELNET:
                fillTelnet(connection);
                break;
        }
    }

    private int selectionIndexFor(ServerType serverType) {
        List<ServerType> serverTypes = getServerTypes();
        for (int i = 0; i < serverTypes.size(); i++) {
            if (serverType == serverTypes.get(i)) return i;
        }
        return -1;
    }

    private void setTextViewContent(int id, String value) {
        setTextViewContent(getView(), id, value);
    }

    private void fillFHEMWEB(FHEMServerSpec connection) {
        View view = getView();
        if (view == null) return;

        setTextViewContent(view, R.id.url, connection.getUrl());
        setTextViewContent(view, R.id.username, connection.getUsername() + "");
        setTextViewContent(view, R.id.password, connection.getPassword());
        setTextViewContent(view, R.id.clientCertificatePath, connection.getClientCertificatePath());
        setTextViewContent(view, R.id.serverCertificatePath, connection.getServerCertificatePath());
        setTextViewContent(view, R.id.clientCertificatePassword, connection.getClientCertificatePassword());

        CheckBox clientCertificateCheckbox = (CheckBox) getView().findViewById(R.id.enableCertificateAuthentication);
        clientCertificateCheckbox.setChecked(connection.isClientCertificateEnabled());
    }

    private void fillTelnet(FHEMServerSpec connection) {
        View view = getView();
        if (view == null) return;

        setTextViewContent(view, R.id.ip, connection.getIp());
        setTextViewContent(view, R.id.port, connection.getPort() + "");
        setTextViewContent(view, R.id.password, connection.getPassword());
    }

    private void setTextViewContent(View view, int id, String value) {
        if (view == null) return;

        TextView textView = (TextView) view.findViewById(id);
        if (textView != null) {
            textView.setText(value);
        }
    }

    private interface ConnectionTypeDetailChangedListener {
        void onChanged();
    }
}

File: src/main/java/li/klass/fhem/fragments/connection/ConnectionDetailFragment.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.fragments.connection

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.method.PasswordTransformationMethod
import android.view.*
import android.widget.*
import com.github.angads25.filepicker.model.DialogConfigs
import com.github.angads25.filepicker.model.DialogProperties
import com.github.angads25.filepicker.view.FilePickerDialog
import com.google.common.base.Preconditions.checkNotNull
import com.google.common.collect.Lists.newArrayList
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import li.klass.fhem.R
import li.klass.fhem.constants.Actions
import li.klass.fhem.constants.BundleExtraKeys
import li.klass.fhem.dagger.ApplicationComponent
import li.klass.fhem.fhem.connection.FHEMServerSpec
import li.klass.fhem.fhem.connection.ServerType
import li.klass.fhem.fragments.core.BaseFragment
import li.klass.fhem.service.connection.ConnectionService
import li.klass.fhem.util.PermissionUtil
import org.jetbrains.anko.coroutines.experimental.bg
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*
import javax.inject.Inject

class ConnectionDetailFragment : BaseFragment() {
    private var connectionId: String? = null
    private var isModify = false
    private var connectionType: ServerType? = null
    private var detailChangedListener: ConnectionTypeDetailChangedListener? = null

    @Inject
    lateinit var connectionService: ConnectionService

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun setArguments(args: Bundle) {
        super.setArguments(args)
        if (args.containsKey(BundleExtraKeys.CONNECTION_ID)) {
            connectionId = args.getString(BundleExtraKeys.CONNECTION_ID)
            isModify = true
        }
    }

    override fun inject(applicationComponent: ApplicationComponent) {
        applicationComponent.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        var view = super.onCreateView(inflater, container, savedInstanceState)
        if (view != null) {
            return view
        }

        view = inflater!!.inflate(R.layout.connection_detail, container, false)

        val connectionTypeSpinner = view!!.findViewById(R.id.connectionType) as Spinner
        if (isModify) {
            connectionTypeSpinner.isEnabled = false
        }

        val connectionTypes = serverTypes

        val adapter = ArrayAdapter(activity,
                android.R.layout.simple_spinner_dropdown_item, connectionTypes)
        connectionTypeSpinner.adapter = adapter

        connectionTypeSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, position: Int, id: Long) {
                handleConnectionTypeChange(connectionTypes[position])
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {}
        }

        return view
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.connection_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item!!.itemId == R.id.save) {
            handleSave()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun mayPullToRefresh(): Boolean = false

    private val serverTypes: List<ServerType>
        get() {
            val connectionTypes = newArrayList<ServerType>()
            connectionTypes.addAll(Arrays.asList(*ServerType.values()))
            connectionTypes.remove(ServerType.DUMMY)
            return connectionTypes
        }

    @SuppressLint("InflateParams")
    private fun handleConnectionTypeChange(connectionType: ServerType) {
        if (view == null) return

        this.connectionType = checkNotNull(connectionType)
        val activity = checkNotNull(activity)
        val inflater = checkNotNull(activity.layoutInflater)

        val view: View?
        when (connectionType) {
            ServerType.FHEMWEB -> {
                view = inflater.inflate(R.layout.connection_fhemweb, null)
                handleFHEMWEBView(view)
            }
            ServerType.TELNET -> view = inflater.inflate(R.layout.connection_telnet, null)
            else -> throw IllegalArgumentException("cannot handle connection type " + connectionType)
        }

        assert(view != null)

        val showPasswordCheckbox = view!!.findViewById(R.id.showPasswordCheckbox) as CheckBox?
        val passwordView = view.findViewById(R.id.password) as EditText?
        if (showPasswordCheckbox != null && passwordView != null) {
            showPasswordCheckbox.setOnClickListener { myView ->
                val radio = myView as CheckBox
                val checked = radio.isChecked
                if (checked) {
                    passwordView.transformationMethod = null
                } else {
                    passwordView.transformationMethod = PasswordTransformationMethod.getInstance()
                }
            }

            if (isModify) showPasswordCheckbox.isEnabled = false
        }

        val showCertificatePasswordCheckbox = view.findViewById(R.id.showCertificatePasswordCheckbox) as CheckBox?
        val passwordClientCertificateView = view.findViewById(R.id.clientCertificatePassword) as EditText?
        if (showCertificatePasswordCheckbox != null && passwordClientCertificateView != null) {
            showCertificatePasswordCheckbox.setOnClickListener { myView ->
                val radio = myView as CheckBox
                val checked = radio.isChecked
                if (checked) {
                    passwordClientCertificateView.transformationMethod = null
                } else {
                    passwordClientCertificateView.transformationMethod = PasswordTransformationMethod.getInstance()
                }
            }

            if (isModify) showCertificatePasswordCheckbox.isEnabled = false
        }

        val connectionPreferences = checkNotNull<View>(getView()).findViewById(R.id.connectionPreferences) as ViewGroup
        checkNotNull(connectionPreferences)
        connectionPreferences.removeAllViews()
        connectionPreferences.addView(view)

        if (detailChangedListener != null) detailChangedListener!!.onChanged()
    }

    private fun handleSave() {
        view ?: return

        val saveStrategy = strategyFor(connectionType ?: ServerType.FHEMWEB)
        val saveData = saveStrategy.saveDataFor(view!!) ?: return

        val myContext = context
        async(UI) {
            bg {
                if (isModify) {
                    connectionService.update(connectionId!!, saveData, myContext)
                } else {
                    connectionService.create(saveData, myContext)
                }
            }.await()
            activity.sendBroadcast(Intent(Actions.BACK))
        }
    }

    private fun strategyFor(connectionType: ServerType?): ConnectionStrategy {
        val type = connectionType ?: ServerType.FHEMWEB
        return when (type) {
            ServerType.FHEMWEB -> FhemWebStrategy(context)
            ServerType.TELNET -> TelnetStrategy(context)
            else -> throw IllegalArgumentException("don't know what to do with " + type)
        }
    }

    private fun handleFHEMWEBView(view: View?) {
        val setClientCertificate = view!!.findViewById(R.id.setClientCertificatePath) as ImageButton
        setClientCertificate.setOnClickListener(View.OnClickListener {
            if (getView() == null) return@OnClickListener

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                PermissionUtil.checkPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE)
            }
            val clientCertificatePath = getView()!!.findViewById(R.id.clientCertificatePath) as TextView
            val initialPath = File(clientCertificatePath.text.toString())

            val properties = DialogProperties()
            properties.selection_mode = DialogConfigs.SINGLE_MODE
            properties.selection_type = DialogConfigs.FILE_SELECT
            properties.root = initialPath

            val dialog = FilePickerDialog(activity, properties)
            dialog.setTitle(R.string.selectFile)
            dialog.setDialogSelectionListener { files ->
                if (files.isNotEmpty()) {
                    clientCertificatePath.text = files[0]
                }
            }
            dialog.show()
        })
    }

    override fun getTitle(context: Context): CharSequence? =
            context.getString(R.string.connectionManageTitle)

    override fun update(refresh: Boolean) {
        if (!isModify) {
            LOG.info("I can only update if a connection is being modified!")
            activity.sendBroadcast(Intent(Actions.DISMISS_EXECUTING_DIALOG))
            return
        }

        async(UI) {
            val result = bg {
                connectionService.forId(connectionId!!, getContext())
            }.await()
            if (result == null) {
                LOG.error("update - cannot find server with ID $connectionId")
                activity.sendBroadcast(Intent(Actions.BACK))
            } else {
                setValuesForCurrentConnection(result)
            }
        }
    }

    private fun setValuesForCurrentConnection(connection: FHEMServerSpec) {
        val view = view ?: return

        // We do not need to change the type selector here, as the right one is already selected.
        // We just overwrite values within the edit fields.
        if (connection.serverType == connectionType) {
            fillDetail(connection)
        } else {
            // We have to change the detail view to the one which is right for the current
            // connection type. However, we do not know when the selection changed listener
            // of the combo box fires. This is why we register a global listener, which is called
            // when the new view has been attached to the root view.
            // Afterwards we can continue with filling the fields with the respective values
            // of the current connection!
            detailChangedListener = object : ConnectionTypeDetailChangedListener {
                override fun onChanged() {
                    detailChangedListener = null

                    fillDetail(connection)
                }
            }
            val connectionTypeSpinner = view.findViewById(R.id.connectionType) as Spinner
            connectionTypeSpinner.setSelection(selectionIndexFor(connection.serverType), true)
        }
    }

    private fun fillDetail(fhemServerSpec: FHEMServerSpec) {
        view ?: return
        strategyFor(connectionType).fillView(view!!, fhemServerSpec)
    }

    private fun selectionIndexFor(serverType: ServerType): Int {
        val serverTypes = serverTypes
        return serverTypes.indices.firstOrNull { serverType == serverTypes[it] } ?: -1
    }

    private interface ConnectionTypeDetailChangedListener {
        fun onChanged()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(ConnectionDetailFragment::class.java)
    }
}
